### PR TITLE
Copy of "[Test][CI][gfx90a] Basic support for gfx90a: Jenkinsfile, dockerfile, unit tests (#1082)"

### PR DESCRIPTION
**Pull request by junliume**
copy of # 1082

### CODE REVIEW

**Review by asroy — APPROVED** [review]

File: Jenkinsfile [hunk 1]
```diff
@@ -273,6 +275,7 @@ pipeline {
         Full_test       = " -DMIOPEN_TEST_ALL=On"
         Tensile_build_env = "MIOPEN_DEBUG_HIP_KERNELS=0 "
         Tensile_setup = " -DMIOPEN_TEST_MIOTENSILE=ON -DMIOPEN_USE_MIOPENTENSILE=ON -DMIOPEN_USE_ROCBLAS=OFF"
+        gfx90a_flags = "-DMIOPEN_TEST_GFX90a=ON"
```
Comment (shurale-nkn) [inline Jenkinsfile:278]: Remove this flag. Cmake should use auto detection

File: Jenkinsfile [hunk 2]
```diff
@@ -548,6 +578,20 @@ pipeline {
                         buildHipClangJobAndReboot( setup_flags: Tensile_setup + extra_log_env + Bf16_flags, build_env: Tensile_build_env + extra_log_env, test_flags: ' --verbose ', gpu_arch: "gfx908:xnack-", miotensile_version: Tensile_version, target_id: "ON")
                     }
                 }
+                // stage('Fp32 Hip Tensile-Latest All gfx90a') {
```
Comment (shurale-nkn) [inline Jenkinsfile:581]: What about these stages?

File: Jenkinsfile [hunk 3]
```diff
@@ -600,6 +656,10 @@ pipeline {
                 }
                 stage('Fp32 OpenCL All gfx1030') {
                     agent{ label rocmnode("navi21") }
+                    options {
+                        timeout(time: 90, unit: 'MINUTES')
```
Comment (shurale-nkn) [inline Jenkinsfile:660]: why is this required here?

File: Jenkinsfile [hunk 4]
```diff
@@ -610,6 +670,12 @@ pipeline {
                         buildHipClangJobAndReboot(setup_flags: Full_test + Fp16_flags, build_env: WORKAROUND_iGemm_936, build_install: "true", gpu_arch: "gfx908")
                     }
                 }
+                stage('Fp16 Hip All Install gfx90a') {
+                    agent{ label rocmnode("gfx90a") }
+                    steps{
+                        buildHipClangJobAndReboot(setup_flags: Full_test + Fp16_flags + gfx90a_flags, build_env: WORKAROUND_iGemm_936, build_install: "true", gpu_arch: "gfx90a:xnack-")
```
Comment (shurale-nkn) [inline Jenkinsfile:676]: remove gfx90a_flags

File: Jenkinsfile [hunk 5]
```diff
@@ -373,6 +382,16 @@ pipeline {
...
                     }
                 }
+                stage('Fp32 Hip Debug gfx90a /opt/rocm') {
+                    agent{ label rocmnode("gfx90a") }
+                    environment{
+                        gpu_arch = "gfx90a:xnack-"
+                        prefixpath = "/opt/rocm"
+                    }
+                    steps{
+                        buildHipClangJobAndReboot(setup_flags: gfx90a_flags, prefixpath: prefixpath, build_type: 'debug', config_targets: Smoke_targets, gpu_arch: gpu_arch)
```
Comment (shurale-nkn) [inline Jenkinsfile:392]: remove gfx90a_flags

File: Jenkinsfile [hunk 6]
```diff
@@ -667,6 +733,30 @@ pipeline {
                         buildHipClangJobAndReboot( setup_flags: Tensile_setup + Full_test + extra_log_env + Int8_flags, build_env: Tensile_build_env + extra_log_env, test_flags: ' --verbose ', gpu_arch: "gfx908:xnack-", miotensile_version: Tensile_version, target_id: "ON")
                     }
                 }
+                // stage('Fp32 Hip Tensile All gfx90a') {
+                //     agent{ label rocmnode("gfx90a") }
```
Comment (shurale-nkn) [inline Jenkinsfile:737]: What about these stages? *same

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -34,6 +34,7 @@ option( MIOPEN_TEST_HALF "Test in half mode" OFF )
 option( MIOPEN_TEST_INT8 "Test in int8 mode" OFF )
 option( MIOPEN_TEST_BFLOAT16 "Test in bfloat16 mode" OFF )
 option( MIOPEN_TEST_GFX908 "Test on MI100 (gfx908)" OFF )
+option( MIOPEN_TEST_GFX90a "Test on gfx90a" ${MIOPEN_TEST_GFX90a} )
```
Comment (shurale-nkn) [inline test/CMakeLists.txt:37]: auto detection

**Review by shurale-nkn — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 1 above)
Comment (junliume) [reply Jenkinsfile:278]: Autodetection is only possible when #1099 is fully checked in.
Currently all CI nodes are using ROCm4.2, where rocminfo cannot recognize gfx90a properly thus causing GPU Detection to fail.
The "workaround" in this PR is to identify GPU type from arch ID, and pull ROCm4.3 only if gfx90a is set.
Sort of chicken and egg paradox: we need ROCm4.3 to detect gfx90a properly and then pull ROCm4.3 to build the docker :)
#1099 is going to fix the problem. However, since upgrading the ROCm version for all CI might be a big change, I hope to check in this one to unlock a lot of gfx90a related PRs. 
Thanks!

File: Jenkinsfile (hunk 2 above)
Comment (junliume) [reply Jenkinsfile:581]: They are tracked by this issue:
https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1094

File: Jenkinsfile (hunk 3 above)
Comment (junliume) [reply Jenkinsfile:660]: Fp32 OpenCL All gfx1030 recently has some stability issues: ref: https://github.com/ROCmSoftwarePlatform/MIOpen/issues/285

Since it is OCL only and likely not a bug caused by MIOpen changes:
1.  Set a retry(2) for this stage ONLY. because if it failed before, the only way is to restart the whole set of parallel stages, which will take hours and many resources.
2. timeout is for an observation that when this stage is "unstable", usually it will *stuck* after the failure for hours (last observation is that it has ctest failure but stuck for 3 hours). We want it to fail fast and with minimal manual intervention.

File: Jenkinsfile (hunk 4 above)
Comment (junliume) [reply Jenkinsfile:676]: Same as here: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1098#discussion_r692316081

File: Jenkinsfile (hunk 5 above)
Comment (junliume) [reply Jenkinsfile:392]: Same as here: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1098#discussion_r692316081

File: test/CMakeLists.txt (hunk 7 above)
Comment (junliume) [reply test/CMakeLists.txt:37]: Same as here :): https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1098#discussion_r692316081

File: Jenkinsfile (hunk 6 above)
Comment (junliume) [reply Jenkinsfile:737]: Same as here :) : https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1098#discussion_r692316833

File: Jenkinsfile (hunk 1 above)
Comment (shurale-nkn) [reply Jenkinsfile:278]: but wait. This flag used only in 4 stages, but gfx90a was tested in 7 stages. So, the 3 stages work with autodetection and work fine.

Comment (shurale-nkn) [reply Jenkinsfile:278]: For this PR ROCm4.3 required only on gfx90a, docker rocm selection based on value in 'gpu_arch'. 
So no need to use (gfx90a_flags = "-DMIOPEN_TEST_GFX90a=ON") 
As I can see from the logs, the GPU was identified correctly.

Comment (junliume) [reply Jenkinsfile:278]: Yes, you are right :) gpu_arch and gfx90a_flags are redundant checks.
Let me make the changes and remove this flag. Thanks! @shurale-nkn

File: test/CMakeLists.txt (hunk 7 above)
Comment (shurale-nkn) [reply test/CMakeLists.txt]: same

**Review by shurale-nkn — APPROVED** [review]
LGTM

**Review by JehandadKhan — APPROVED** [review]

**Review by shurale-nkn — APPROVED** [review]
